File: security-application/src/main/java/com/example/security/application/port/in/ListAlertsUseCase.java
package com.example.security.application.port.in;

import com.example.security.domain.common.TenantId;
import com.example.security.domain.rule.Alert;
import com.example.security.domain.rule.Alert.AlertStatus;
import java.time.Instant;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * use case 6 — 운영자가 발화된 알람 timeline 조회.
 *
 * <p>tenantId 강제, status 필터, 기간 필터, cursor pagination.
 */
public interface ListAlertsUseCase {

  Page list(ListAlertsQuery query, OperatorContext operator);

  Alert acknowledge(UUID alertId, OperatorContext operator);

  Alert resolve(UUID alertId, OperatorContext operator);

  Alert markFalsePositive(UUID alertId, OperatorContext operator);

  record ListAlertsQuery(
      TenantId tenantId,
      Optional<AlertStatus> status,
      Optional<Instant> from,
      Optional<Instant> to,
      int size,
      Optional<UUID> after) {

    /** 알람 timeline 한 페이지 상한 — UI 가 cursor 로 추가 페이지 요청. (API4 — 자원 소비 제한) */
    public static final int MAX_SIZE = 500;

    public ListAlertsQuery {
      if (size < 1 || size > MAX_SIZE) {
        throw new IllegalArgumentException("size 는 1~" + MAX_SIZE + ": " + size);
      }
    }
  }

  record Page(List<Alert> alerts, UUID nextCursor) {}
}

File: security-application/src/main/kotlin/com/example/security/application/port/in/ListAlertsUseCase.kt
package com.example.security.application.port.`in`

import com.example.security.domain.common.TenantId
import com.example.security.domain.rule.Alert
import java.time.Instant
import java.util.Optional
import java.util.UUID

/**
 * use case 6 — 운영자가 발화된 알람 timeline 조회.
 *
 * tenantId 강제, status 필터, 기간 필터, cursor pagination.
 */
interface ListAlertsUseCase {

    fun list(query: ListAlertsQuery, operator: OperatorContext): Page

    fun acknowledge(alertId: UUID, operator: OperatorContext): Alert

    fun resolve(alertId: UUID, operator: OperatorContext): Alert

    fun markFalsePositive(alertId: UUID, operator: OperatorContext): Alert

    @JvmRecord
    data class ListAlertsQuery(
        val tenantId: TenantId,
        val status: Optional<Alert.AlertStatus>,
        val from: Optional<Instant>,
        val to: Optional<Instant>,
        val size: Int,
        val after: Optional<UUID>,
    ) {

        init {
            require(size in 1..MAX_SIZE) { "size 는 1~$MAX_SIZE: $size" }
        }

        companion object {
            /** 알람 timeline 한 페이지 상한 — UI 가 cursor 로 추가 페이지 요청. (API4 — 자원 소비 제한) */
            // const val 로 노출 — Java 의 @Max(ListAlertsQuery.MAX_SIZE) 가 컴파일 타임 상수를 요구.
            const val MAX_SIZE: Int = 500
        }
    }

    @JvmRecord
    data class Page(val alerts: List<Alert>, val nextCursor: UUID?)
}
